File: android/app/src/main/java/com/bhojanos/orderbhojan/nativehost/NativeTrackCohort.kt
package com.bhojanos.orderbhojan.nativehost

import android.content.Context

/**
 * Cohort: internal email allowlist and/or sticky % bucket.
 * Mirror of JS evaluateNativeTrackRollout (native double-check).
 */
object NativeTrackCohort {
    fun inCohort(
        context: Context,
        userEmail: String?,
        deviceId: String?,
    ): Boolean {
        val email = userEmail?.trim()?.lowercase().orEmpty()
        if (email.isNotEmpty() && NativeFeatureFlags.internalEmails(context).contains(email)) {
            return true
        }
        val percent = NativeFeatureFlags.trackPercent(context)
        if (percent <= 0) return false
        if (percent >= 100) return true
        val key = deviceId?.trim().orEmpty().ifEmpty { "anonymous" }
        val bucket = stickyBucket0to99("native-track|$key")
        return bucket < percent
    }

    fun shouldOpenNative(context: Context, userEmail: String?, deviceId: String?): Boolean {
        return NativeFeatureFlags.isNativeHostEnabled(context) &&
            NativeFeatureFlags.isNativeTrackEnabled(context) &&
            inCohort(context, userEmail, deviceId)
    }

    fun stickyBucket0to99(cohortKey: String): Int {
        var hash = 0x811c9dc5.toInt()
        for (ch in cohortKey) {
            hash = hash xor ch.code
            hash *= 0x01000193
        }
        val unsigned = hash.toLong() and 0xffffffffL
        return (unsigned % 100L).toInt()
    }
}

File: ios/App/App/NativeTrackCohort.swift
import Foundation

enum NativeTrackCohort {
    static func inCohort(userEmail: String?, deviceId: String?) -> Bool {
        let email = (userEmail ?? "").trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        if !email.isEmpty && NativeFeatureFlags.internalEmails.contains(email) {
            return true
        }
        let percent = NativeFeatureFlags.trackPercent
        if percent <= 0 { return false }
        if percent >= 100 { return true }
        let key = (deviceId ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        let material = key.isEmpty ? "anonymous" : key
        let bucket = stickyBucket0to99("native-track|\(material)")
        return bucket < percent
    }

    static func shouldOpenNative(userEmail: String?, deviceId: String?) -> Bool {
        NativeFeatureFlags.isNativeHostEnabled
            && NativeFeatureFlags.isNativeTrackEnabled
            && inCohort(userEmail: userEmail, deviceId: deviceId)
    }

    static func stickyBucket0to99(_ cohortKey: String) -> Int {
        var hash: UInt32 = 0x811c9dc5
        for byte in cohortKey.utf8 {
            hash ^= UInt32(byte)
            hash = hash &* 0x01000193
        }
        return Int(hash % 100)
    }
}
